tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -2391,6 +2638,161 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  FillInputs(transpose_to_nhwc, &transpose_nhwc_control_edges,
+             &transpose_nhwc_in);
+
+  const int mklop_num_inputs = mklop->num_inputs();
+  gtl::InlinedVector<Node*, 4> mklop_control_edges;
+  gtl::InlinedVector<std::pair<Node*, int>, 4> mklop_in(mklop_num_inputs);
+  FillInputs(mklop, &mklop_control_edges, &mklop_in);
+
+  const int transpose_nchw_num_inputs = transpose_to_nchw->num_inputs();
+  gtl::InlinedVector<Node*, 4> transpose_nchw_control_edges;
+  gtl::InlinedVector<std::pair<Node*, int>, 4> transpose_nchw_in(
+      transpose_nchw_num_inputs);
+  FillInputs(transpose_to_nchw, &transpose_nchw_control_edges,
+             &transpose_nchw_in);
+
+  // We use same name as original node, but change the op
+  // name.
+  NodeBuilder nb(mklop->name(), mklop->type_string());
+
+  for (int i = 0; i < mklop_num_inputs; i++) {
+    if (mklop_in[i].first == transpose_to_nhwc) {
+      // Fill "x":
+      nb.Input(transpose_nhwc_in[0].first, transpose_nhwc_in[0].second);
+    } else {
+      // Fill inputs other than "x":
+      nb.Input(mklop_in[i].first, mklop_in[i].second);
+    }
+  }
+
+  copy_attrs(const_cast<const Node*>(mklop), &nb, true);
+  nb.Attr("data_format", data_format);
+
+  // Copy the device assigned to old node to new node.
+  nb.Device(mklop->def().device());
+
+  // Create node.
+  Node* new_node;
+  TF_CHECK_OK(nb.Finalize(&**g, &new_node));
+  CHECK_NOTNULL(new_node);
+
+  // Fill outputs.
+  for (const Edge* e : transpose_to_nchw->out_edges()) {
+    if (!e->IsControlEdge()) {
+      const int kTransposeWithMklOpOutputSlot = 0;
+      CHECK_NOTNULL((*g)->AddEdge(new_node, kTransposeWithMklOpOutputSlot, e->dst(),
+                                  e->dst_input()));
+    }
+  }
+
+  // Copy device assigned to old node to new node.
+  new_node->set_assigned_device_name(mklop->assigned_device_name());
+
+  // Copy requested_device and assigned_device_name_index
+  new_node->set_requested_device(mklop->requested_device());
+  new_node->set_assigned_device_name_index(mklop->assigned_device_name_index());
+
+  (*g)->RemoveNode(transpose_to_nhwc);
+  (*g)->RemoveNode(mklop);
+  (*g)->RemoveNode(transpose_to_nchw);
+
+  return Status::OK();
+}
+
+Status MklLayoutRewritePass::FuseNode(
+    std::unique_ptr<Graph>* g, std::vector<Node*>& nodes,
+    const MklLayoutRewritePass::FusionInfo fi) {
+  return fi.fuse_func(g, nodes, fi.copy_attrs);
+}
+
+std::tuple<bool, std::vector<Node*>, const MklLayoutRewritePass::FusionInfo>
+MklLayoutRewritePass::CheckForNodeFusion(Node* a) const {
+  const FusionInfo* fi_ptr = nullptr;
+
+  for (auto fi = finfo_.begin(); fi != finfo_.end(); ++fi) {
+    fi_ptr = &*fi;
+    //
+    // Make sure node "a" and its succeding nodes (b, c ...), match the pattern
+    // defined in fusion info (ops[0], ops[1], ...),
+    // aka. "a->b->c" matches "op1->op2->op3"
+    //
+
+    std::stack<Node *, std::vector<Node *>> work_stack;
+    std::unordered_set<Node *> visited_nodes;
```

Comment:
Using a global `visited_nodes` set is not enough. We can use the same graph as last time for example: fusion pattern `A->B->C` with graph `A->C`, `A->B`, and `B->C`. `C` is visited first and marked as a `visited` node, so by the time we reach `B`, we won't go to `C` anymore. (This case seems relevant enough. Please also add a test case for this.)

I also don't like that we have to iterate pass the already processed `current_node->out_edges()` every time (even though all we do with them is a quick check). Instead of a `visited_nodes` set, you can just save the current out edge for each checker index, `0 <= index < node_checkers.size()`. The next time you revisit this checker index level you can just go to the next edge. You can also get rid of the `work_stack` and store the results directly in `nodes` to avoid having to copy it at the end. For example:
```c++
std::vector</*EdgeSet iterator type*/> current_neighbor(fi->node_checkers.size());
// To get the candidate node inside the loop
candidate_node = current_neighbor[index];
// To have the next candidate ready
++current_neighbor[index];
// To store the candidate node.
nodes[index] = candidate_node;
// To check if you're out of neighbor
if (current_neighbor[index] == nodes[index-1]->out_edges().end())
```

Reply:
You're right, that's a good counter-example. My algorithm has implications that subgraph is always a tree (whose root is "A"), that may not be true in real-world models. 

I modified the code, applied the idea you provide that "remove the loop for out_edges()" and "record edge iterator instead of visited nodes", but I probably not fully understand your idea:
1. I don't think we can totally get rid of the `work_stack`, because overall this is a stack-based algorithm and we need a stack to help us roll back, that's what `work_stack` does.
2. I prefer to to have `current_neighbor` as a `stack` rather than `vector`, then it will expand and shrink just as what `work_stack` does at the same time. This could avoid handling the `index` when rollback, it's more natural I think.

Reply:
Anyway, I commit the change just now. To give us more confidence, I wrote 3 test cases:
1. A->B->C;                    // The most naive case, single input with single output.
2. A->B->C; A->B->D;   // Need the capability to roll back.
3. A->B->C; A->C;         // Handle the non-tree case.

And I confirmed these 3 patterns could successfully recognized in my implementation now. Please see my code for details.